torch/fx/experimental/symbolic_shapes.py

```diff
@@ -254,10 +254,49 @@ def _sympystr(self, printer):
 
         @classmethod
         def eval(cls, base, divisor):
+            # NOTE [ Checking types with SymPy ]
+            # Python has a dedicated complex type, but SymPy represents complex
+            # with Add exprs. So we cannot provide the same error message here
+            # as in Python floordiv because there's no easy way to distinguish
+            # an arbitrary Add expr from a complex number.
+            #
+            # isinstance doesn't work for arbitrary exprs and for Symbols, even
+            # when the latter have integer=True set during construction. So you
+            # are supposed to use the is_* properties, but it's also not always
+            # reliable as 0 has both is_integer and is_real set to True no
+            # matter whether it's constructed as a Float or an Integer.
+            #
+            # Also, booleans cannot be used in arithmetic exprs in SymPy and
+            # cannot be passed to the Integer and Float constructors.
+            # https://github.com/pytorch/pytorch/issues/90900
+            def check_supported_type(x):
+                # Note: is_real returns True for both Integer and Float, but
+                # let's be explicit here so this reflects the error message
+                # below.
+                if not x.is_integer and not x.is_real:
+                    raise TypeError(
+                        f"unsupported operand type(s) for //: "
+                        f"'{type(base).__name__}' and '{type(divisor).__name__}'"
+                        f", expected integer or real")
```

Comment:
OK. What I expect to see is that if we have bugs in assumption propagation, we will incorrectly fire TypeError. I think the correct fix is to get assumption propagation fixed. But since testing for complex inputs is very niche, I think it would also be OK to delete this logic if necessary.